### ISSUE TITLE
Invalid links returned with some chinese characters as delimiters

### ISSUE THREAD

**Issue by tommedema** [markdown-it/linkify-it#15]
Steps to reproduce:

1. linkify the following text

【视频奇志大兵《发烧友》 在线观看 - 酷6视频】奇志大兵《发烧友》 在线观看，奇志 大兵 搞笑双簧 _ 发烧友 （追星族）  http://t.cn/RZwjG7U（分享自 @酷6网）

2. the output link is

`http://t.cn/RZwjG7U（分享自`

whereas the output link should be

`http://t.cn/RZwjG7U`

The reason is that （ is not recognized as a separating delimiter, yet it is quite common in Chinese.

Out of 500 posts I gathered, about 20 of them had links like this, resulting in invalid links reported by linkify.

**Comment by puzrin** [comment]
Could you post a permalink to demo http://markdown-it.github.io/linkify-it/ ? You can type there all examples at once, with results.

**Comment by puzrin** [comment]
[Example](http://markdown-it.github.io/linkify-it/#t1=http%3A%2F%2Ft.cn%2FRZwjG7U%EF%BC%88%E5%88%86%E4%BA%AB%E8%87%AA%20%40%E9%85%B76%E7%BD%91%EF%BC%89%0A)

Problem is in `(`, used in your example. There are NO space before it. I'm not familiar with asian puncctuation. Are such `(` always word terminators? Is is possible to have it in links, like in wiki?

**Comment by tommedema** [comment]
See http://markdown-it.github.io/linkify-it/#t1=http%3A%2F%2Ft.cn%2FRv4VRqQ%EF%BC%88%E5%88%86%E4%BA%AB%E8%87%AA%20%40%E7%88%B1%E5%A5%87%E8%89%BA%EF%BC%89%0A%0Ahttp%3A%2F%2Fhttp%3A%2F%2Ft.cn%2FR2t287H

I added another issue where a user mistakenly added http:// twice.

These are both (technically speaking) user behavior issues, but the first one especially is quite common.

Formally speaking ( is not a word terminator, however because it takes so much white space, many chinese people don't put a space in front of it (it's an act of being lazy, but it's quite common).

Perhaps I should do some string manipulation before I pass it to `linkify-it`?

**Comment by puzrin** [comment]
Example with double http:// is technically correct. That's like `http://localhost` - any local domain allowed. Imho such typo do not need fixes.

Example with `(` is more serious. It uses chineese unicode scopes, and can be distinguished from english scopes. But i need recommendations about grammar to be sure.

**Comment by tommedema** [comment]
I'm sorry I'm not a Chinese speaker. Perhaps someone else can help out here. There are also other Chinese punctuation characters, like `，`

**Comment by puzrin** [comment]
Let's leave it open until someone can formalize info for asian languages group (probably japan language has the same issues). I'm ready to fix as soon as possible, but would like to avoid kludges with inccomplete workarounds.

**Comment by tommedema** [comment]
More examples of how difficult these Chinese posts are to parse: http://markdown-it.github.io/linkify-it/

What's your idea on this?

**Comment by puzrin** [comment]
Seems you given wrong link, it shows default text.

**Comment by tommedema** [comment]
Sorry, I've fixed the link now.

**Comment by puzrin** [comment]
Thanks for examples. Posted a question at commonmark forum http://talk.commonmark.org/t/linkifier-lets-discuss-and-test/1045/9?u=vitaly

As far as i see in last example, spaces are not used at all. It's possible to to track locale change (non-english -> english), but that's not safe.

**Comment by tommedema** [comment]
That's correct, spaces are not used. The Chinese use different punctuation marks or even don't use them because URLs are in English script and therefore the disctinction is easily visible to the eye, but more difficult for machines.

**Comment by puzrin** [comment]
What about links with chineese chars? I can find link start somehow by `[ any non english ]http://`, but can't use this rule to search link end.

PS. Anyway, it worth to start collecting test samples for china language separately. Something like [this](https://github.com/markdown-it/linkify-it/tree/master/test/fixtures)

**Comment by tommedema** [comment]
You're absolutely right. I think the first rule makes sense, but URLs can probably have chinese characters. Perhaps the first rule by itself would help though, and somekind of smart processing for certain delimiters that are unlikely to appear within an URL, e.g.:

`str.replace(/[\uff08\uff09\uff0c\uff01\u3002\uff1f\u3010\u3011\uff3b\uff3d\u3001\u300a\u300b\u2605]/g, ' ') //（ ） ， ！ 。 ？ 【 】 ［ ］ 、 《 》 ★`

**Comment by liuyang1204** [comment]
Found this URL doesn't match:

http://markdown-it.github.io/linkify-it/#t1=http%3A%2F%2F172.26.142.48%2Fviewerjs%2F%23..%2F0529%2Fslides.pdf

BTW, I found a website comparing different regexps for URL:
https://mathiasbynens.be/demo/url-regex

**Comment by liuyang1204** [comment]
I made new issue #18 for this problem

**Comment by tommedema** [comment]
By the way, Facebook does this properly, but I'm not sure if they use a proprietary solution.

**Comment by fengmk2** [comment]
I will try to resolve this problem.

**Comment by puzrin** [comment]
@fengmk2 implementation can be not easy, but for the first step it would be enougth to have collection of fixtures with good coverage of Chinese edge cases.

See https://github.com/markdown-it/linkify-it/tree/master/test/fixtures

**Comment by mikelambert** [comment]
Here is my example page linkifying incorrectly:
http://www.dancedeets.com/events/896736620379772/2016-r16-taiwan-x-wbc

In this case, it's a vertical bar, though to be precise it's `｜` and not `|` (I hope you can see the difference properly). The page author uses it as a delimiter, but linkify sees it as part of the domain name (presumably it sees it as just another word character).

**Comment by puzrin** [comment]
@mikelambert Nothing to fix. Fuzzy mode is not safe, it's an author's mistale to use linkify-it in wrong way.

**Comment by mikelambert** [comment]
I'm the website author using linkify-it to add links to raw text from a variety of sources which I didn't write myself (facebook events and other websites). And the individual source authors are not using linkify-it, and of course not writing their text with linkify-it in mind.

I recognize that fuzzy is not safe, and not perfect. However, it seems unfortunate that a vertical bar (what some authors are using as visual punctuation) is treated as part of a domain name. It seemed like the fuzziness could be smarter, even if it's still fuzzy.

So to clarify, is this a "it's not a bug, just user error" bug, or a "it's not a bug I care about fixing, but patches are welcome" bug?

**Comment by puzrin** [comment]
@mikelambert I mean, problems with `|` in that examples is because linkifier applied after layout compose, instead of before compose. But linkifier goal is to find links in natural texts, not everywhere (that's impossible).

My personal opinion is, that linkifier is used not as expected. So, this example is not good enougth as a reason for changes. May be i don't understand something, but this is my opinion for now. If i had to do such site, i would parse links first, then compose header.

**Comment by mikelambert** [comment]
What do you mean by "layout compose" ? I assume you are referring to writing up the text and adding the `|` to lay things out?

https://www.facebook.com/events/896736620379772/ is the source text I am working with. (Notice that Facebook gets the linkification correct.) I receive the raw text from the FB API, and then am trying to linkify it for use on my own website.

I understand that linkifier-it might be the wrong tool (since this is not natural text), and I am using it incorrectly (since it is applied after the author's layout composition). But unfortunately, I am not the author of the text, and so I am not able to linkify *before* adding the `|` characters.

Thanks for your time!

**Comment by puzrin** [comment]
Then, if you have source in known format (`|`-separated), i would split it first, then apply linkifier, then join back. May be, with some additional conditions (line should be short and have chineese letters inside). Or would try to get html if possible (not familiar with FB api)

The reason to add `|` support could be, if you say "humans write such way" or "that's de-facto standard", with a lot of examples.

**Comment by puzrin** [comment]
In other words, if you have auto-generated text - consider parse/detect it's structure prior to apply linkifier. Or if you know some new uncovered de-facto patterns of human writing - create new issue with proofs (live examples), and i'll try to fix it if possible.

**Comment by mikelambert** [comment]
Thank you, that's a creative solution to this problem. I'll go with that for now.

I'll create a separate issue with the few examples I have, and you can decide if it's justified "de-facto pattern of human writin" or not there. :)

**Comment by episodeyang** [comment]
@mikelambert I'm wondering what does facebook linkify do with this following link?

link: `https://zh.wikipedia.org/wiki/（   `

**Comment by mikelambert** [comment]
Feel free to create a dummy facebook event (or even a FB post on your wall) and see what happens?

It seems like it fails to parse that link properly, instead linking to https://zh.wikipedia.org/wiki/ .

**Comment by showonne** [comment]
hi, I found this problem still persists to this day. Why not use ` ） ， ！ 。 ？ 【 】 ［ ］ 、 《 》 ★` as separators ?

**Comment by puzrin** [comment]
Please use english and make sure to keep focus on this concrete issue.

**Comment by WangJincheng4869** [comment]
> Please use english and make sure to keep focus on this concrete issue.

由于就算使用英文也无法与你进行正常的沟通，因为你只会反复重复那句话：“请给出真实的使用实例”。

真给了你又不看，既然这样我为什么还要选择使用英文？尊重是相互的，你不配。

**Comment by hlwz5735** [comment]
@WangJincheng4869 我非常同意你的观点，哥们，理论上非常简单的事儿，真的，哪那么麻烦啊！

求人不如求己，还是选择自己打补丁吧：

```typescript
import type * as LinkifyIt from 'linkify-it'

// 非法 URL 字符，主要是中文标点符号
const INVALID_URL_CHARS = /[，。、！？：；（）【】《》〉」』〕〗]/

const urlValidator: LinkifyIt.Validate = (text: string, pos: number, self: any) => {
  const slice = text.slice(pos)

  // 1. 延迟编译 http 正则（与 linkify-it 内部逻辑保持一致）
  // 由于我们覆盖了默认的 http: schema，需要自己来创建这个正则
  if (!self.re.http) {
    self.re.http = new RegExp(
      '^\\/\\/' + self.re.src_auth + self.re.src_host_port_strict + self.re.src_path,
      'i',
    )
  }

  const match = slice.match(self.re.http)

  if (!match) {
    return 0
  }

  const fullMatch = match[0]

  // 2. 从左向右扫描，遇到第一个非法字符就截断（你的原始逻辑）
  for (let i = 0; i < fullMatch.length; i++) {
    if (INVALID_URL_CHARS.test(fullMatch[i])) {
      // 返回截断后的长度
      return i
    }
  }

  return fullMatch.length
}

const schemas: Record<string, unknown> = {
  'http:': { validate: urlValidator },
  'https:': { validate: urlValidator },
}
```

**Comment by puzrin** [comment]
I tried to fix CJK punctuation at least, but found this collision

```
https://zh.wikipedia.org/wiki/（
https://zh.wikipedia.org/wiki/（你让我）神魂颠倒
https://zh.wikipedia.org/wiki/（後）大中華佛國
https://zh.wikipedia.org/wiki/（我得不到）满足
https://zh.wikipedia.org/wiki/（步）解
https://zh.wikipedia.org/wiki/（海員）社會保障公約（修訂本）
https://zh.wikipedia.org/wiki/（） 

https://zh.wikipedia.org/wiki/《
https://zh.wikipedia.org/wiki/《127小时》获奖名单
https://zh.wikipedia.org/wiki/《1961年教育法令》
https://zh.wikipedia.org/wiki/《2025年監獄（修訂）規則》
https://zh.wikipedia.org/wiki/《5.16通知》
https://zh.wikipedia.org/wiki/《BanG_Dream!》手游

https://ja.wikipedia.org/wiki/（
https://ja.wikipedia.org/wiki/（ハル）
https://ja.wikipedia.org/wiki/（タイトルのないアルバム）
https://ja.wikipedia.org/wiki/（株）
https://ja.wikipedia.org/wiki/（株）悪の秘密結社
https://ja.wikipedia.org/wiki/（特）情報とってもインサイト
```

Wiki really uses CJK punctuation in links. And it's not clear how to make changes without false positives.

Unfortunately, "easy fix" failed.

**Comment by geyang** [comment]
Can just use AI!

**Comment by hlwz5735** [comment]
> I tried to fix CJK punctuation at least, but found this collision
> 
> ```
> https://zh.wikipedia.org/wiki/（
> https://zh.wikipedia.org/wiki/（你让我）神魂颠倒
> https://zh.wikipedia.org/wiki/（後）大中華佛國
> https://zh.wikipedia.org/wiki/（我得不到）满足
> https://zh.wikipedia.org/wiki/（步）解
> https://zh.wikipedia.org/wiki/（海員）社會保障公約（修訂本）
> https://zh.wikipedia.org/wiki/（） 
> 
> https://zh.wikipedia.org/wiki/《
> https://zh.wikipedia.org/wiki/《127小时》获奖名单
> https://zh.wikipedia.org/wiki/《1961年教育法令》
> https://zh.wikipedia.org/wiki/《2025年監獄（修訂）規則》
> https://zh.wikipedia.org/wiki/《5.16通知》
> https://zh.wikipedia.org/wiki/《BanG_Dream!》手游
> 
> https://ja.wikipedia.org/wiki/（
> https://ja.wikipedia.org/wiki/（ハル）
> https://ja.wikipedia.org/wiki/（タイトルのないアルバム）
> https://ja.wikipedia.org/wiki/（株）
> https://ja.wikipedia.org/wiki/（株）悪の秘密結社
> https://ja.wikipedia.org/wiki/（特）情報とってもインサイト
> ```
> 
> Wiki really uses CJK punctuation in links. And it's not clear how to make changes without false positives.
> 
> Unfortunately, "easy fix" failed.

When identifying hyperlinks, if a slash is followed by a punctuation mark, truncate the link early (before that punctuation).

Maybe you're not a native Chinese speaker, which is why you're splitting hairs over this.

The result you just demonstrated with your example is exactly what I (and my native Chinese-speaking friends) want. In Chinese convention, any URL containing special characters should be encoded with encodeURI(). If you don't encode it and get a broken result — you deserve it, and users completely understand that.

Wiki has a special case with parentheses at the beginning of a link, but that's just an exception. The "easy fix" I proposed works for 99% of real-world cases. Even if it fails on those rare exceptions, users get it.

But the current implementation looks ridiculous and counterintuitive to native Chinese speakers. Sacrificing proper linking in over 90% of everyday scenarios just to handle a <1% edge case — one that Chinese users already consider malformed — is exactly why I say the author is splitting hairs.

The author is technically right that parentheses can cause some unsolvable ambiguity. That's a sharp observation, and it's true. But what the author probably doesn't realize is that Chinese users don't really care about that ambiguity — what they do care about is normal punctuation being pointlessly included in links. So I still believe the "easy fix" makes sense. I'm confident the vast majority of native Chinese speakers agree with me.

**Comment by puzrin** [comment]
@hlwz5735 , you probably misunderstand the nuances of how heuristic algorithms work.  Chinese links do not work not because they are poorly done, but because no appropriate rule exists at all.

The difficulties with adding rule is that it should not produce false positives. That's not a personal preference, but the only way to keep such algorithms maintainable without endless bug fixes. => This requires a list of desired cases, usage frequencies, and error probabilities. I don't know CJK and don't visit CJK sites to collect such info with adequate quality.

If you have time to work carefully with patterns and stats, we can do this in a separate issue to stay focused. In the worst case, I could try to organize CJK support as a plugin with a comment "use at your own risk".

**Comment by hlwz5735** [comment]
Thanks for the detailed explanation. You've made it clear that you care more about theoretical correctness and algorithm maintainability than about real-world usability for Chinese users. That's your choice as the maintainer.

I disagree, but this discussion is going nowhere. We're operating on completely different premises — yours being "perfect algorithm first", mine being "works for real people". You think your way is the only right way. I think that's exactly what being a nerd means in this context. We're approaching this from completely different perspectives — you from pure engineering, me from daily usability.

So let's just end the discussion here. We're not going to convince each other. I'll continue using my own "easy fix" for my own needs, and you can keep your current implementation. No hard feelings, but this is my last reply.

**Comment by tats-u** [comment]
> ```
> https://zh.wikipedia.org/wiki/（
> https://zh.wikipedia.org/wiki/（你让我）神魂颠倒
> https://zh.wikipedia.org/wiki/（後）大中華佛國
> https://zh.wikipedia.org/wiki/（我得不到）满足
> https://zh.wikipedia.org/wiki/（步）解
> https://zh.wikipedia.org/wiki/（海員）社會保障公約（修訂本）
> https://zh.wikipedia.org/wiki/（） 
> 
> https://zh.wikipedia.org/wiki/《
> https://zh.wikipedia.org/wiki/《127小时》获奖名单
> https://zh.wikipedia.org/wiki/《1961年教育法令》
> https://zh.wikipedia.org/wiki/《2025年監獄（修訂）規則》
> https://zh.wikipedia.org/wiki/《5.16通知》
> https://zh.wikipedia.org/wiki/《BanG_Dream!》手游
> 
> https://ja.wikipedia.org/wiki/（
> https://ja.wikipedia.org/wiki/（ハル）
> https://ja.wikipedia.org/wiki/（タイトルのないアルバム）
> https://ja.wikipedia.org/wiki/（株）
> https://ja.wikipedia.org/wiki/（株）悪の秘密結社
> https://ja.wikipedia.org/wiki/（特）情報とってもインサイト
> ```

Many of these examples could be resolved by extending the next section of the GFM specification, such as using [`Bidi_Paired_Brackets`](https://www.unicode.org/Public/UCD/latest/ucd/BidiBrackets.txt):


> When an autolink ends in `)`, we scan the entire autolink for the total number of parentheses. If there is a greater number of closing parentheses than opening ones, we don’t consider the unmatched trailing parentheses part of the autolink, in order to facilitate including an autolink inside a parenthesis:
>
> [Example 625](https://github.github.com/gfm/#example-625)
>
> ```
> www.google.com/search?q=Markup+(business)
>
> www.google.com/search?q=Markup+(business)))
>
> (www.google.com/search?q=Markup+(business))
>
> (www.google.com/search?q=Markup+(business)
> ```
>
> ```
> <p><a href="http://www.google.com/search?q=Markup+(business)">www.google.com/search?q=Markup+(business)</a></p>
> <p><a href="http://www.google.com/search?q=Markup+(business)">www.google.com/search?q=Markup+(business)</a>))</p>
> <p>(<a href="http://www.google.com/search?q=Markup+(business)">www.google.com/search?q=Markup+(business)</a>)</p>
> <p>(<a href="http://www.google.com/search?q=Markup+(business)">www.google.com/search?q=Markup+(business)</a></p>
> ```

**Comment by tats-u** [comment]
Non-ASCII characters in the above URLs are percent-encoded when copied from the browser's address bar to the clipboard. Additionally, in Chrome, internationalized domain names are converted to Punycode when copied.

https://はじめよう.みんな/テスト

→ https://はじめよう.みんな/%E3%83%86%E3%82%B9%E3%83%88 (Firefox)
→ https://xn--p8j9a0d9c9a.xn--q9jyb4c/%E3%83%86%E3%82%B9%E3%83%88 (Chrome)

**Comment by tats-u** [comment]
> `http://t.cn/RZwjG7U（分享自 @酷6网）`

Hmm, it's impossible to perfectly distinguish boundaries in all cases. We have no choice but to make some compromises somewhere.

**Comment by puzrin** [comment]
> > `http://t.cn/RZwjG7U（分享自 @酷6网）`
> 
> Hmm, it's impossible to perfectly distinguish boundaries in all cases. We have no choice but to make some compromises somewhere.

@tats-u 

The problem is HOW to organize that compromise. Because false positives in heuristic approaches are very unwelcome. I could organize regxps to factory class, for example. And slice into smaller fragments. That should simplify creating plugins.

The second problem - lack of case statistics. People post examples and thoughts, but without statistics, they cannot be landed in the mainstream without risk.

- If you have time to process CJK link stats more seriously, I would suggest to start a new issue, where we can be more focused (this one is already too long).
- As an alternative, you can tell me which fragments of RE should be easy to modify, and I will prepare api for your plugin. Keeping uncertain things in plugin should solve the problem of unwanted collisions.
 
---

In short:

- Direct CJK landing to mainstream can be a very boring process, requires a dedicated trusted person, with good knowledge of links used in CKJ.
- I still believe a quick-fix is possible via plugins (after api improvement). As a temporary/research solution.

**Comment by puzrin** [comment]
> Many of these examples could be resolved by extending the next section of the GFM specification, such as using [`Bidi_Paired_Brackets`](https://www.unicode.org/Public/UCD/latest/ucd/BidiBrackets.txt):

I would be happy if people followed this convention. But I can not promote the "new standard". I can only work with "what we have now". Creating the spec is a separate, difficult task. I'd prefer to stay on the implementation level in this concrete situation. From my side, I can help to make experiments easy. IMHO, that's constructive and honest.

**Comment by qq1023535906** [comment]
> [@WangJincheng4869](https://github.com/WangJincheng4869) 我非常同意你的观点，哥们，理论上非常简单的事儿，真的，哪那么麻烦啊！
> 
> 求人不如求己，还是选择自己打补丁吧：
> 
> import type * as LinkifyIt from 'linkify-it'
> 
> // 非法 URL 字符，主要是中文标点符号
> const INVALID_URL_CHARS = /[，。、！？：；（）【】《》〉」』〕〗]/
> 
> const urlValidator: LinkifyIt.Validate = (text: string, pos: number, self: any) => {
>   const slice = text.slice(pos)
> 
>   // 1. 延迟编译 http 正则（与 linkify-it 内部逻辑保持一致）
>   // 由于我们覆盖了默认的 http: schema，需要自己来创建这个正则
>   if (!self.re.http) {
>     self.re.http = new RegExp(
>       '^\\/\\/' + self.re.src_auth + self.re.src_host_port_strict + self.re.src_path,
>       'i',
>     )
>   }
> 
>   const match = slice.match(self.re.http)
> 
>   if (!match) {
>     return 0
>   }
> 
>   const fullMatch = match[0]
> 
>   // 2. 从左向右扫描，遇到第一个非法字符就截断（你的原始逻辑）
>   for (let i = 0; i < fullMatch.length; i++) {
>     if (INVALID_URL_CHARS.test(fullMatch[i])) {
>       // 返回截断后的长度
>       return i
>     }
>   }
> 
>   return fullMatch.length
> }
> 
> const schemas: Record<string, unknown> = {
>   'http:': { validate: urlValidator },
>   'https:': { validate: urlValidator },
> }

感谢哥们给出解决方案，我对 markdown-it 的使用不了解，你这个是从自定义插件角度解决的吗？

我当时只是想着从上游根本性的解决这个问题，于是创建了一个 issues 来阐述我的需求。作者说要一些资料，我就花了一些时间整理了一下。我想的是虽然我写代码菜，但是我可以帮忙整理些资料，也算是有些贡献。

但是这哥们就是钻牛角尖，油盐不进啊。也不知道是语言不通导致的，还是这哥们对技术的追求导致的。我深知我是白嫖党，所以也算是把姿态压的很低。但是士可杀不可辱啊，最终没压制住就来吐槽了下，随后就把我拉黑了。

我在使用过程中发现 markdown-it 还是太重了，最终用了腾讯开源的那个项目 cherry-markdown。这个还对流式输入进行了专门的优化，一举两得了，即提升了性能，也解决了 url 识别不准的问题。


最后：

@puzrin 对不起，哥们，当时是我没控制好情绪，我为此道歉！

**Comment by puzrin** [comment]
In v6, Unicode punctuation (`{P}` class) is no longer part of the link by default. This should solve most of the problems. Also added an example of paired scopes extending. In v6, modifying rules is much simpler.

Closing this issue. Please create new ones if something significant remains unresolved.
